ruby/computer_science/time_complexity.md

```diff
@@ -0,0 +1,300 @@
+
+To do that we need to delve into a new concept. Big O.
+
+### What is Big O?
+
+Big O gives us a consistent way to measure the efficiency of an algorithm. It gives us a measurement for the time it takes for an algorithm to run as the input grows so that you can directly compare the performance of two algorithms and pick the best one.
+
+Big O is not a piece of code you can put your algorithm into and it tells you how efficient it is. You will need to measure how the number of steps changes as the data grows and using this you can apply a Big O Notation to it and measure it against other algorithms. In many cases you'll be using a data structure in which the ways you interact with it are well known, and in that case it's easier to judge how it will scale as the input changes.
+
+Firstly we'll summarise the Big O Notations and then provide a little more context for each one. The reading materials will dive into greater detail.
+
+#### Big O Notation
+
+The Big O Notations in the order of best scalability to worst are:
+
+- `O(1)` - Constant Complexity
+- `O(log N)` - Logarithmic Complexity
+- `O(N)` - Linear Complexity
+- `O(N log N)` - N x log N Complexity
+- `O(n<sup>2</sup>)` - Quadratic Complexity
+- `O(n<sup>3</sup>)` - Cubic Complexity
+- `O(2<sup>N</sup>)` - Exponential Complexity
+- `O(N!)` - Factorial Complexity
+
+#### O(1) - Constant Complexity
+
+To understand Constant Complexity let's use a simple array.
+
+~~~ruby
+arr = [1, 2, 3, 4, 5]
+~~~
+
+If we want to look up what is at index 2, we can get to the element using `arr[2]` which would give us back `3`. This takes just one step. If we double our array...
+
+~~~ruby
+arr = [1, 2, 3, 4, 5, 6, 7, 8, 9, 10]
+~~~
+
+I can still access any element in just one step. `arr[7]` gives me `8` in a single step. Our array can keep growing and I can always access any element in a single step. It's constant. Hence we have `O(1)`.
+
+Looking up something in one step is as good as it gets for time complexity.
+
+While we're looking at the simplest form of Big O, let's take a look at one of its little gotchas to keep in mind. You may have thought a moment ago, is it really just one step? The answer is technically no, in reality the computer must first look up where the array is in memory, then from the first element in the array it needs to jump to the index argument provided. That's at least a couple of steps. So you wouldn't be wrong for writing something like `O(1 + 2(steps))`. However, the 2 steps are merely incidental. With an array of 10,000 elements it still takes the same amount of steps as if the array was 2 elements. Because of this, Big O doesn't concern itself with these incidental numbers. They don't provide any context to how the complexity grows when the data size changes, because they are constant, and so in Big O they are dropped. Big O only wants to tell us an algorithm's complexity relative to the size of the input.
```

Comment:
Read more on why Big O drops constants?

Reply:
I should've been more specific.

```md
in reality the computer must first look up where the array is in memory, then from the first element in the array it needs to 
jump to the index argument provided. That's at least a couple of steps. So you wouldn't be wrong for writing something 
like `O(1 + 2(steps))`.
```
I'm curious about the computer taking at least two steps to read a value of an array and how Big-O ends up being `O(1+2(steps))` for this operation.

Reply:
> I should've been more specific.
> 
> ```lisp
> in reality the computer must first look up where the array is in memory, then from the first element in the array it needs to 
> jump to the index argument provided. That's at least a couple of steps. So you wouldn't be wrong for writing something 
> like `O(1 + 2(steps))`.
> ```
> 
> I'm curious about the computer taking at least two steps to read a value of an array and how Big-O ends up being `O(1+2(steps))` for this operation.

Ahhh. I read a little bit about it in a common sense guide to data structures and algorithms. Big O is `O(1)` since it's constant time. It's just that technically if you actually counted the steps takes each array look up has some arbitray number of steps first. I don't know if it's actually 2. I guess it depends on the programming language.

I was just trying to illustrate the point that saying something is constant time doesn't mean it's literally a one step operation in programming terms. Just say as the input grows that number of steps doesn't change.